### ISSUE TITLE
"Glue Joint" or "Fixed Joint" (for both 2D and 3D)

### ISSUE THREAD

**Issue by greenfox1505** [godotengine/godot-proposals#86]
**Describe the project you are working on:** It is a 2D Physics Game. I have a "background" layer and a "foreground" layer. Most objects in my game have different collision shapes in each layer. For example, the player almost exclusively exists in the Foreground, however large objects have both Foreground and Background elements. Foreground elements are not passable by the player, but background elements cannot intersect background elements from other objects.

My solution as it stands, I have two independent RigidBodies2D with their own collision shapes and two PinJoint2D with two different positions attached to both nodes. It is quite buggy in edge cases of large objects and strong forces.

**Describe how this feature / enhancement will help your project:** A single joint that locks position and rotation relative to each other. For my game, I need a 2D version, however Godot also lacks a 3D version.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:** Same as a PinJoint.

**Describe implementation detail for your proposal (in code), if possible:** I believe this is just a PinJoint with a relative rotation locking?

**If this enhancement will not be used often, can it be worked around with a few lines of script?:** It can be approximated by creating two PinJoint2D with different positions or three PinJoint (3d) with different translations. However, a single "joint" is a cleaner, hopefully less buggy, approach.

**Is there a reason why this should be core and not an add-on in the asset library?:** It's a low level feature that likely should be part of the physics engine.

Unity calls this a "Fixed Joint", so maybe that is the naming scheme would should follow, but "glue" seemed more descriptive to me.

**Comment by girng** [comment]
> My solution as it stands, I have two independent RigidBodies2D with their own collision shapes and two PinJoint2D with two different positions attached to both nodes. It is quite buggy in edge cases of large objects and strong forces.

Might be a bug with PinJoint2D? If it is, feel free to create an issue about it so it can be fixed =]

**Comment by greenfox1505** [comment]
It's just a rounding error. The fix is to move the 2nd joint further away from the first one. However, it would still be better to have a joint that locks rotation.

**Comment by girng** [comment]
> It's just a rounding error. The fix is to move the 2nd joint further away from the first one. However, it would still be better to have a joint that locks rotation.

I see. Seems like a property could be added to the node to enable this behavior? Would that work?

**Comment by greenfox1505** [comment]
Yeah, that could work. "Match rotation" or "lock relative rotation" flag? Could be useful for the Grove joint too. Although spring joint might want one for each attachment, so a car's socks are locked vertical, but the wheel can still spin.

**Comment by JPTeasdale** [comment]
+1 to this, except you can set the bounds that the rotation is locked.

**Comment by Ughuuu** [comment]
I am interested in this, but saw the code and don't know that much about 2d physics. I could research I guess, as I really want this feature, could look how other physics engines implement it.
But was curious where this discussion was left. So basically right now there is no way for joints to "remain glued" at a specific angle. Should this angle be fixed, should it be a range. Does this complicate the implementation?
In any case, for my use cases and for most use cases out there, if the joint would have a range, both the one where it's glued(having minRange and maxRange equal) and the one where its an interval would work.
Also, this API would be nice if it would contain generic things that are also available in most physics engines out there(box2d, rapier) so that when those get integrated as options it's easy to implement.

If I were to take this, could I just make a PR on godot repo and start a discussion there, or do we continue here the discussion until we reach a conclusion?

**Comment by Marenz** [comment]
Opened a PR implementing this for 2D: godotengine/godot#120424

It adds an optional angular lock to `PinJoint2D` (`angular_lock_enabled`, plus `angular_lock_softness` / `angular_lock_play` tuning). A `PinJoint2D` already locks relative position, so the angular lock turns it into a rigid weld / fixed joint — the two bodies keep their own collision shapes but a fixed relative orientation and move as one unit.

**Comment by Schnorg** [comment]
See comments on joint locking on [Add angular lock (rigid weld / fixed joint) option to PinJoint2D
#120424](https://github.com/godotengine/godot/pull/120424)
